### ISSUE TITLE
about reference of weight init according to layer depth or layer id 

### ISSUE THREAD

**Issue by SeunghyunSEO** [pytorch/torchtitan#375]
hi, first of all, thank you for the nice opensource project !
i just have been reading your model code and found it initialize model weights following num_layers or layer_id.
it is not conventional like kaming init (std=1/\sqrt{fan_in}) or GPT-2 init (std=0.02).
and it also dost not look like MuP or something.
so i just want to know if there are any references or it's just empirical for training stability.

edit) i forgot std=0.02/\sqrt{depth} init for output layers of residual block is from GPT-2 paper. sry ! just wondering where depth_init is from

**Comment by wanchaol** [comment]
@lessw2020

**Comment by lessw2020** [comment]
Hi @SeunghyunSEO  - sorry for the delay, didn't see this earlier.
To your question - the depth init came about from research last summer when we we were doing work on parallel attention blocks. I did a comparison sweep and adding the depth init was the winner, so have continued to use that. 
I'm not sure where the concept came from though - it bubbled up in discussions with IBM research a while back. 
I did see that Olmo was also using this, and they referred to it as a "mitchell init" but I was not able to find anything in arxiv on it. 
Anyway, short answer is it's empirically based.  We haven't done a sweep on it though since llama3 came out so maybe we will revisit it in the future but it continues to perform well in our training runs.
Hope that helps!

**Comment by SeunghyunSEO** [comment]
thank you for the kind answer @lessw2020 !
i guess it makes sense because Mitchell init's layerwise output variance would be more consistent compared to GPT init

**Comment by garrett361** [comment]
I had this same question. Some analysis, in case anyone is interested:

* `depth_init = False`: rescales output `nn.Linear` layers by `~ 1 / sqrt(n_layers)`. This is what gpt2 does: [Sec 2.3](https://cdn.openai.com/better-language-models/language_models_are_unsupervised_multitask_learners.pdf). The idea, presumably, is that the final output from the sum of all residuals is $z \sim \sum_{\ell=0}^{L-1} r_\ell$ for $L$ layers. The sum vanishes in expectation over model param init, $E[z]=0$, so its size is characterized by its second moment, $E[z^2]$ and the $r_\ell \sim 1 / \sqrt{L}$ initialization enforces that $E[z^2] \sim  \mathcal{O}(1)$ is independent of $L$ at leading order.  
* `depth_init = True`: instead rescales by `~ 1 / sqrt(1 + layer_idx)`.   Less clear motivation. Now $E[z^2] \sim \mathcal{O}(\ln L)$. 

 > I was not able to find anything in arxiv on it.

Only thing I could find is this [paper, eq 15.](https://arxiv.org/pdf/1908.11365). [Link to similar code in OLMo here.](https://github.com/allenai/OLMo/blob/090253dac6688f2532509daa7aa2eb5fae50e956/olmo/config.py?plain=1#L204-L209) Couldn't immediately find any examples in `transformers` with `~ 1 / sqrt(1 + layer_idx)`, as a comparison point. Haven't been able to find too much discussion of init schemes in recent tech reports. 

`depth_init = True` is currently the default everywhere. Seems a little questionable given the apparent lack of code and literature support. WDYT @lessw2020  @tianyu-l ? CC @rakkit, maybe of interest.

I think there are some other places where the init defaults should be visited, too. One is the embedding std which is the unusually-large [1.0 for all models](https://github.com/pytorch/torchtitan/blob/30ca37a9cff98bd152e5a2ec36cc6ad11d3b7b6d/torchtitan/models/llama3/__init__.py?plain=1#L68) except gpt-oss, which has the [more-standard 0.02](https://github.com/pytorch/torchtitan/blob/30ca37a9cff98bd152e5a2ec36cc6ad11d3b7b6d/torchtitan/models/gpt_oss/__init__.py?plain=1#L58).

**Comment by tianyu-l** [comment]
@garrett361 we don't have to keep `depth_init` flag, if it's not standard practice in the literature / open recipes. Could you help revisit the init in torchtitan, hopefully with well-cited references? That will be much appreciated!

**Comment by garrett361** [comment]
> it bubbled up in discussions with IBM research a while back

Yes, also asking around IBM about the above

**Comment by rakkit** [comment]
thx @garrett361   we did few depth scaling related ablation on our local torchtitan branch last year (see page 27 https://arxiv.org/abs/2510.03871) 
include combination
on init side
1)identical init
2)scale by relative-depth (i forgot where this comes from)
3)scale by total-depth (gpt style)

on residual / non-residual multipler 
1)identical
2)depth-mup or completeP scale
3)depth-normalized ([moduler-norm ](https://arxiv.org/abs/2405.14813) style)


from what we see the current "depth_init = false" (i.e. gpt style, scale by total number of layers) in torchtitan works best.   but i agree that we should make this depth-wise init and residual multipliers more clear,  set Bool type on "depth_init" is kind confusing

**Comment by garrett361** [comment]
Great input, thanks @rakkit ! Will include this when I write up findings.

**Comment by rakkit** [comment]
sure @garrett361 i will also try to re-run this sweep at recent

**Comment by garrett361** [comment]
> scale by relative-depth (i forgot where this comes from)

Yeah this corresponds to depth_init = True in latest main, right? The scheme that I didn't see a great motivation for. 

> also asking around IBM about the above

I was told that the depth_init = True scheme probably came from considerations for post-norm Transformer architectures that aren't really relevant any more (everything in titan is pre-norm). 

More generally, some mechanism for easily experimenting with custom init schemes and custom per-param lrs in titan would be really useful, but that's a larger issue for another day.

**Comment by tianyu-l** [comment]
> some mechanism for easily experimenting with custom init schemes and custom per-param lrs in titan would be really useful

@garrett361 

I believe https://github.com/pytorch/torchtitan/pull/2633 and following PRs from @fegin have addressed the custom init part.

I'll submit a PR for per-param (group) learning rate and weight-decay soon.

**Comment by garrett361** [comment]
Cool, hadn't seen. I'll check it out

**Comment by wwwjn** [comment]
As #2633 landed, I will temporarily close this PR. Feel free to reopen it if needed @garrett361
